macosx/QuickLookExtension/PreviewProvider.mm

```diff
@@ -0,0 +1,290 @@
+// It may be used under the MIT (SPDX: MIT) license.
+// License text can be found in the licenses/ folder.
+
+#import "PreviewProvider.h"
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <libtransmission/torrent-metainfo.h>
+#include <libtransmission/utils.h>
+
+#import "NSStringAdditions.h"
+
+static NSUInteger const kIconWidth = 16;
+
+namespace
+{
+
+class FileTreeNode
+{
+  public:
+    FileTreeNode() = default;
+    ~FileTreeNode() = default;
+
+    auto MaybeCreateChild(std::string_view child_name)
+    {
+        return children_.try_emplace(std::string{ child_name });
+    }
+
+  private:
+    FileTreeNode(FileTreeNode const&) = delete;
+    FileTreeNode& operator=(FileTreeNode&) = delete;
+
+    std::unordered_map<std::string, FileTreeNode> children_;
+};
+
+} // namespace
+
+NSString* generateIconData(UTType* type, NSUInteger width, NSMutableDictionary<NSString*, QLPreviewReplyAttachment*>* allImgProps)
+{
+    // We need to do this once per file type, per image size
+    NSString* iconFileName = [NSString stringWithFormat:@"%@.%ld.tiff", type.identifier, width];
+
+    if (!allImgProps[iconFileName])
+    {
+        NSImage* icon = [[NSWorkspace sharedWorkspace] iconForContentType:type];
+
+        NSRect const iconFrame = NSMakeRect(0.0, 0.0, width, width);
+        NSImage* renderedIcon = [[NSImage alloc] initWithSize:iconFrame.size];
+        [renderedIcon lockFocus];
+        [icon drawInRect:iconFrame];
+        [renderedIcon unlockFocus];
+
+        NSData* iconData = renderedIcon.TIFFRepresentation;
+
+        QLPreviewReplyAttachment* imageAttachment = [[QLPreviewReplyAttachment alloc] initWithData:iconData contentType:UTTypePNG];
+
+        allImgProps[iconFileName] = imageAttachment;
+    }
+
+    return [@"cid:" stringByAppendingString:iconFileName];
+}
+
+@implementation PreviewProvider
+
+- (void)providePreviewForFileRequest:(QLFilePreviewRequest*)request
+                   completionHandler:(void (^)(QLPreviewReply* _Nullable reply, NSError* _Nullable error))handler
+{
+    QLPreviewReply* reply = [[QLPreviewReply alloc]
+        initWithDataOfContentType:UTTypeHTML
+                      contentSize:CGSizeMake(1200, 800)
+                dataCreationBlock:^NSData* _Nullable(QLPreviewReply* _Nonnull replyToUpdate, NSError* __autoreleasing _Nullable* _Nullable error) {
+                    NSString* previewHTML = [self generateHTMLPreviewFor:request.fileURL andReply:replyToUpdate];
+
+                    return [previewHTML dataUsingEncoding:NSUTF8StringEncoding];
+                }];
+
+    handler(reply, nil);
+}
+
+- (NSString*)generateHTMLPreviewFor:(NSURL*)url andReply:(QLPreviewReply* _Nonnull)replyToUpdate
+{
+    // Try to parse the torrent file
+    auto metainfo = tr_torrent_metainfo{};
+    if (!metainfo.parse_torrent_file(url.path.UTF8String))
+    {
+        return nil;
+    }
+
+    NSBundle* bundle = [NSBundle mainBundle];
```

Comment:
Ditto about property syntax

Reply:
Done in 37b22630c1217893d5e0da6bab7272f7acfcbc6e